### PR TITLE
rgw/admin: add an interface around the HTTP Client and mocking

### PR DESCRIPTION
**Pull request by leseb**
Partially solve https://github.com/ceph/go-ceph/issues/508.
I'm saying "partially" since @phlogistonjohn we need to see if we keep `Debug` or not.

## Checklist
- [ ] Added tests for features and functional changes
- [ ] Public functions and types are documented
- [ ] Standard formatting is applied to Go code

### CODE REVIEW

File: rgw/admin/radosgw.go [hunk 1]
```diff
@@ -28,19 +28,24 @@ var (
 	errNoSecretKey = errors.New("secret key not set")
 )
 
+// HTTPClient interface that conform to the http package
```
Comment (phlogistonjohn) [inline rgw/admin/radosgw.go:31]:
```suggestion
// HTTPClient interface that conforms to that of the http package's Client.
```

**Review by phlogistonjohn — COMMENTED** [review]
Changes look pretty good so far.
As for "Debug or not" I don't think we need to retain it for backwards compatibility as the whole of rgw/admin was already preliminary/alpha state. Personally, I would prefer we drop Debug and the "built in" dependency on logging in rgw/admin. But it doesn't need to all be done in this PR FWIW - I'd be happy to approve this and change (or discuss changing) Debug in a new pr.

File: rgw/admin/radosgw.go [hunk 2]
```diff
@@ -28,19 +28,24 @@ var (
...
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
 type verbHTTP string
```
Comment (thotz) [inline rgw/admin/radosgw.go:36]: Not related to this PR, but this can be removed since https://github.com/ceph/go-ceph/pull/531/ got merged right?

**Review by thotz — DISMISSED** [review]
This PR LGTM, I have comment but related to #531. Sorry it got missed in earlier PR review

**Comment by leseb** [comment]
> Changes look pretty good so far.
> As for "Debug or not" I don't think we need to retain it for backwards compatibility as the whole of rgw/admin was already preliminary/alpha state. Personally, I would prefer we drop Debug and the "built in" dependency on logging in rgw/admin. But it doesn't need to all be done in this PR FWIW - I'd be happy to approve this and change (or discuss changing) Debug in a new pr.

Yes, it'd be nice to discuss a bit how/when to remove the Debug entry in another PR. Thanks!

File: rgw/admin/user_test.go [hunk 3]
```diff
@@ -135,3 +138,24 @@ func (suite *RadosGWTestSuite) TestUser() {
 		assert.NoError(suite.T(), err)
 	})
 }
+
+func TestGetUserMockAPI(t *testing.T) {
+	r := ioutil.NopCloser(bytes.NewReader(fakeUserResponse))
+	mockClient := &MockClient{
+		MockDo: func(req *http.Request) (*http.Response, error) {
+			if req.URL.RawQuery == "format=json&uid=dashboard-admin" && req.Method == "GET" && req.URL.Path == "127.0.0.1/admin/user" {
```
Comment (phlogistonjohn) [inline rgw/admin/user_test.go:146]: Possibly a holdover from older code but should "GET" be swapped to a constant from the http package?

**Review by phlogistonjohn — APPROVED** [review]
One very small nitpick. One that I don't see as a blocking issue so I'm approving.

File: rgw/admin/user_test.go (hunk 3 above)
Comment (leseb) [reply rgw/admin/user_test.go:146]: This was introduced with the new interface and the mocking. I can change that.

**Comment by leseb** [comment]
@Mergifyio rebase

**Review by phlogistonjohn — APPROVED** [review]
LGTM

File: rgw/admin/mock.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package admin
+
+import (
+	"net/http"
+)
+
+// MockClient is the mock of the HTTP Client
+// It can be used to mock HTTP request/response from the rgw admin ops API
+type MockClient struct {
```
Comment (ansiwen) [inline rgw/admin/mock.go:9]: I don't see a reason to export this, since the tests are run from inside the admin package.

File: rgw/admin/mock.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	// MockDo is a type that mock the Do method from the HTTP package
+	MockDo MockDoType
```
Comment (ansiwen) [inline rgw/admin/mock.go:11]: dito

File: rgw/admin/mock.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+// MockDoType is a custom type that allows setting the function that our Mock Do func will run instead
+type MockDoType func(req *http.Request) (*http.Response, error)
```
Comment (ansiwen) [inline rgw/admin/mock.go:15]: dito

**Review by ansiwen — COMMENTED** [review]
I don't see a good reason to export the mock, otherwise LGTM

**Review by thotz — APPROVED** [review]
LGTM

**Comment by thotz** [comment]
@ansiwen I guess my approval is treated as second approval and gives `mergify bot` to merge the PR. Apologies for the trouble, hence requesting @leseb please address @ansiwen comments as separate PR

File: rgw/admin/mock.go (hunk 4 above)
Comment (leseb) [reply rgw/admin/mock.go:9]: What about consumers of the library? Should they write their own `MockClient`?

Comment (ansiwen) [reply rgw/admin/mock.go:9]: Well, they have to anyways, right? What we have here is just a wrapper around a function pointer. The function itself must be implemented in any case. The problem I see is that if you export something, it becomes part of the contract with the consumers, and you can't change it anymore. That's exactly what the `HTTPClient` interface is for in this case. And "writing" a `MockClient` is as easy as
```go
type MockDo func(req *http.Request) (*http.Response, error)

func (do MockDo) Do(req *http.Request) (*http.Response, error) { return do(req) }
```
So, I think it's good (for a library) to practice export hygiene and only export what is necessary. I actually would even move this code into the test file.

File: rgw/admin/mock.go (hunk 6 above)
Comment (ansiwen) [reply rgw/admin/mock.go:15]: this doesn't have to be exported in any case, even if you want to offer a MockClient

**Comment by ansiwen** [comment]
> @ansiwen I guess my approval is treated as second approval and gives `mergify bot` to merge the PR. Apologies for the trouble, hence requesting @leseb please address @ansiwen comments as separate PR

Well, it was my own fault, because I didn't select "request changes". 🤷‍♂️

File: rgw/admin/mock.go (hunk 4 above)
Comment (ansiwen) [reply rgw/admin/mock.go:9]: Refernce: https://github.com/golang/go/wiki/CodeReviewComments#interfaces

Comment (leseb) [reply rgw/admin/mock.go:9]: https://github.com/ceph/go-ceph/pull/539

**Pull request merged**